plugins/processors/regex/regex.go

```diff
@@ -0,0 +1,105 @@
+package regex
+
+import (
+	"regexp"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/processors"
+)
+
+type Regex struct {
+	Tags   []converter
+	Fields []converter
+}
+
+type converter struct {
+	Key         string
+	Pattern     string
+	Replacement string
+	ResultKey   string
+}
+
+const sampleConfig = `
+  ## Tag and field conversions defined in a separate sub-tables
+  # [[processors.regex.tags]]
+  #   ## Tag to change
+  #   key = "resp_code"
+  #   ## Regular expression to match on a tag value
+  #   pattern = "^(\\d)\\d\\d$"
+  #   ## Pattern for constructing a new value (${1} represents first subgroup)
+  #   replacement = "${1}xx"
+
+  # [[processors.regex.fields]]
+  #   key = "request"
+  #   ## All the power of the Go regular expressions available here
+  #   ## For example, named subgroups
+  #   pattern = "^/api(?P<method>/[\\w/]+)\\S*" 
+  #   replacement = "${method}"
+  #   ## If result_key is present, a new field will be created
+  #   ## instead of changing existing field
+  #   result_key = "method"
+
+  ## Multiple conversions may be applied for one field sequentially
+  ## Let's extract one more value
+  # [[processors.regex.fields]]
+  #   key = "request"
+  #   pattern = ".*category=(\\w+).*"
+  #   replacement = "${1}"
+  #   result_key = "search_category"
+`
+
+func (r *Regex) SampleConfig() string {
+	return sampleConfig
+}
+
+func (r *Regex) Description() string {
+	return "Transforms tag and field values with regex pattern"
+}
+
+func (r *Regex) Apply(in ...telegraf.Metric) []telegraf.Metric {
+	for _, metric := range in {
+		for _, converter := range r.Tags {
+			if value, ok := metric.Tags()[converter.Key]; ok {
+				metric.AddTag(
+					getKey(converter),
+					getValue(converter, value),
+				)
+			}
+		}
+
+		for _, converter := range r.Fields {
+			if value, ok := metric.Fields()[converter.Key]; ok {
+				switch value := value.(type) {
+				case string:
+					metric.AddField(
+						getKey(converter),
+						getValue(converter, value),
+					)
+				}
+			}
+		}
+	}
+
+	return in
+}
+
+func getKey(c converter) string {
+	if c.ResultKey != "" {
+		return c.ResultKey
+	}
+	return c.Key
+}
+
+func getValue(c converter, value string) string {
+	regex := regexp.MustCompile(c.Pattern)
```

Comment:
We should compile the regex pattern only once on startup, you could do this in Apply with a boolean field on the Regex type so it is only executed once.